File: python/sentinel.py
import sqlite3
import tkinter as tk
import datetime

# Function to calculate Legion Number based on CAC
def resolute(phNum):
    digits = [int(char) for char in phNum]
    while len(digits) > 1:
        digits = [int(char) for char in str(sum(digits))]
    return digits[0]

class Legion:
    def __init__(self, name, score):
        self.name = name
        self.score = score

class SentinelApp:
    def __init__(self):
        self.cac = ""
        self.act = False
        self.db = None
        self.results = None
        self.winningLegions = []
        self.initialize()

    def initialize(self):
        self.winningLegions = []
        self.openDatabaseConnection()
        self.getLegionScores()

    def openDatabaseConnection(self):
        path = "legion_scores.db"
        self.db = sqlite3.connect(path)
        cursor = self.db.cursor()
        cursor.execute("CREATE TABLE IF NOT EXISTS scores (legion_name TEXT, score INTEGER)")
        self.db.commit()

    def getLegionScores(self):
        legionNames = [
            "Legion 1",
            "Legion 2",
            "Legion 3",
            "Legion 4",
            "Legion 5",
            "Legion 6",
            "Legion 7",
            "Legion 8",
            "Legion 9",
        ]
        self.results = []

        cursor = self.db.cursor()
        for legionName in legionNames:
            cursor.execute(
                "SELECT score FROM scores WHERE legion_name = ?",
                (legionName,),
            )
            result = cursor.fetchone()
            score = result[0] if result else 0
            self.results.append(Legion(legionName, score))

    def updateLegionScore(self, legionName, score):
        cursor = self.db.cursor()
        cursor.execute("UPDATE scores SET score = ? WHERE legion_name = ?", (score, legionName,))
        self.db.commit()

    def handleCACInput(self, input):
        self.cac = input.strip()

    def handleAOKInput(self, input):
        self.act = input.strip().lower() == "yes"

    def submitForm(self):
        userLegion = resolute(self.cac)
        currentScore = self.results[userLegion - 1].score
        newScore = currentScore + 1 if self.act else currentScore
        self.updateLegionScore(self.results[userLegion - 1].name, newScore)
        self.getLegionScores()

        # Update the displayed scores
        self.score_label.config(text=self.getLegionScoresText())

        self.determineWinningLegions()

        # Clear the input fields
        self.cac_entry.delete(0, tk.END)
        self.aok_entry.delete(0, tk.END)

        # Update the displayed scores
        self.score_label.config(text=self.getLegionScoresText())

    def determineWinningLegions(self):
        highestScore = max(legion.score for legion in self.results)
        self.winningLegions = [legion for legion in self.results if legion.score == highestScore]

    def getLegionScoresText(self):
        scores_text = "Legion Scores:\n"
        for legion in self.results:
            scores_text += f"{legion.name}: {legion.score}\n"
        return scores_text

    def run(self):
        # Create the GUI window
        root = tk.Tk()
        root.title("Sentinel App")

        # Function to handle the submit button click event
        def submit_form():
            cac = self.cac_entry.get()
            self.handleCACInput(cac)

            aok = self.aok_entry.get()
            self.handleAOKInput(aok)

            self.submitForm()

        # Create and position the input fields
        cac_label = tk.Label(root, text="Enter your City Area Code, e.g. 416 for Toronto:")
        cac_label.pack()
        self.cac_entry = tk.Entry(root)
        self.cac_entry.pack()

        aok_label = tk.Label(root, text='Report an Act Of Kindness by you via typing "yes":')
        aok_label.pack()
        self.aok_entry = tk.Entry(root)
        self.aok_entry.pack()

        # Create and position the submit button
        submit_button = tk.Button(root, text="Submit", command=submit_form)
        submit_button.pack()

        # Create and position the score display label
        self.score_label = tk.Label(root, text=self.getLegionScoresText())
        self.score_label.pack()

        # Create and position the results message label
        current_time = datetime.datetime.now().time()
        if current_time.hour < 23 or (current_time.hour == 23 and current_time.minute < 59):
            results_message = "The leading Legion(s) by 23:59 Hours will take the following day off from Acts Of Kindness"
        else:
            yesterday_winners = ", ".join([legion.name for legion in self.winningLegions])
            results_message = f"Today Legion(s) {yesterday_winners} have a day off from Acts Of Kindness"
        self.results_label = tk.Label(root, text=results_message, fg="red", font=("Arial", 12, "bold"))
        self.results_label.pack()

        # Run the GUI main loop
        root.mainloop()

if __name__ == "__main__":
    app = SentinelApp()
    app.run()

File: +6_QUOTE/Backup/x5/sentinel.js
// JavaScript equivalent of resolute function
function resolute(phNum) {
    let digits = Array.from(String(phNum), Number);
    while (digits.length > 1) {
      digits = Array.from(String(digits.reduce((a, b) => a + b)), Number);
    }
    return digits[0];
  }
  
  class Legion {
    constructor(name, score) {
      this.name = name;
      this.score = score;
    }
  }
  
  class SentinelApp {
    constructor() {
      this.cac = "";
      this.act = false;
      this.db = null;
      this.results = null;
      this.winningLegions = [];
      this.initialize();
    }
  
    initialize() {
      this.winningLegions = [];
      this.openDatabaseConnection();
      this.getLegionScores();
    }
  
    openDatabaseConnection() {
      // No need to connect to SQLite database in JavaScript
    }
  
    getLegionScores() {
      const legionNames = [
        "Legion 1",
        "Legion 2",
        "Legion 3",
        "Legion 4",
        "Legion 5",
        "Legion 6",
        "Legion 7",
        "Legion 8",
        "Legion 9",
      ];
      this.results = [];
  
      for (const legionName of legionNames) {
        const score = 0; // Set initial score to 0 in JavaScript
        this.results.push(new Legion(legionName, score));
      }
    }
  
    updateLegionScore(legionName, score) {
      // No need to update SQLite database in JavaScript
    }
  
    handleCACInput(input) {
      this.cac = input.trim();
    }
  
    handleAOKInput(input) {
      this.act = input.trim().toLowerCase() === "yes";
    }
  
    submitForm() {
      const userLegion = resolute(this.cac);
      const currentScore = this.results[userLegion - 1].score;
      const newScore = this.act ? currentScore + 1 : currentScore;
      this.results[userLegion - 1].score = newScore;
      this.getLegionScores();
  
      // Update the displayed scores
      const scoreLabel = document.getElementById("score-label");
      scoreLabel.innerText = this.getLegionScoresText();
  
      this.determineWinningLegions();
  
      // Clear the input fields
      const cacEntry = document.getElementById("cac-input");
      const aokEntry = document.getElementById("aok-input");
      cacEntry.value = "";
      aokEntry.value = "";
  
      // Update the displayed scores
      scoreLabel.innerText = this.getLegionScoresText();
    }
  
    determineWinningLegions() {
      const highestScore = Math.max(...this.results.map((legion) => legion.score));
      this.winningLegions = this.results.filter((legion) => legion.score === highestScore);
    }
  
    getLegionScoresText() {
      let scoresText = "Legion Scores:\n";
      for (const legion of this.results) {
        scoresText += `${legion.name}: ${legion.score}\n`;
      }
      return scoresText;
    }
  
    run() {
      // Function to handle the submit button click event
      const submitForm = () => {
        const cacInput = document.getElementById("cac-input");
        const aokInput = document.getElementById("aok-input");
        this.handleCACInput(cacInput.value);
        this.handleAOKInput(aokInput.value);
        this.submitForm();
      };
  
      // Add event listener to submit button
      const submitButton = document.getElementById("submit-button");
      submitButton.addEventListener("click", submitForm);
  
      // Create and position the score display label
      const scoreLabel = document.createElement("div");
      scoreLabel.id = "score-label";
      scoreLabel.textContent = this.getLegionScoresText();
      document.body.appendChild(scoreLabel);
  
      // Run the JavaScript code
      this.getLegionScores();
    }
  }
  
  // Create an instance of the SentinelApp class and run it
  const app = new SentinelApp();
  app.run();
  